### PR TITLE
initial setup for running operations in a worker

### PR DESCRIPTION
**Pull request by RebeccaStevens**

### CODE REVIEW

**Comment by AnthorNet** [comment]
Makes me feel like your adding a lot of complexity for very small gains. The amount of "big" operation that will benefit from this is very small, mostly because the most blocking part is the redraw and that cannot be done in workers.
And please no TS :D

**Comment by RebeccaStevens** [comment]
When working on large save games, I've noticed that many operations can take quite awhile to complete. There are several things I want to improve about how operations works which should improve their performance and also improve code readability and reusability.

One of the main things I want change is to [separate concerns](https://en.wikipedia.org/wiki/Separation_of_concerns#:~:text=In%20computer%20science%2C%20separation%20of,code%20of%20a%20computer%20program). Which with regard to operations, that means decoupling their logic from the UI logic and setting up both the batch and single object operations to perform just like one another.

I figured that making these changes would require a lot of refactoring, so I may as well set up operations to use Web Workers for best performance.

I know this is going to take quite a lot of work (and possibly by the time I've finished everything I want to do, I may have rewritten the entire code base 😛) but I'm quite enjoying the work. I've got nothing but free time at the moment so I need something to keep me occupied.

Also note: This PR doesn't contain any actual TypeScript. It just contains [TypeScript declaration files](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) which I'm referencing in [JSDoc comments](https://jsdoc.app/index.html). This give us TypeScript types in regular old JavaScript (though it's not quite as powerful).

**Comment by RebeccaStevens** [comment]
I'll keep this as a draft for now as I might change how things are setup a bit.

**Comment by AnthorNet** [comment]
To be totally honest it's kind of hard to follow, and I don't really wish to get a full new code base...
It makes it hard for me to work on it and on a day to day basis that will not pleasant at all!

I'm still trying to fix the previous commits and still don't see all the advantages.
Beside working a little faster, it also consumes more memory and fails in some larger test saves.
And some layers are still not working properly.

I'm not really convinced yet, and that really bothers me right now to feel that way about the project.

**Comment by RebeccaStevens** [comment]
If it's all good with you I'll just keep doing my thing until everything is in a much nicer state.

I kind of jumped off the deep end into rebuilding core stuff that requires everything else to change as it all depends on that core stuff. I'll keep make this changes on my end and figuring out how things would best fit together. I'll make sure to keep everything well documented.

I'll try to submit small PRs going forward that just change a single small part of the code in a way that works with the existing code and sets it up nicely to work with my larger changes. Hopefully I can manage this so that on your end, you are just getting small PR that are easily manageable.
